pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,449 @@
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            *args : Variable length argument list for objective function
+            **kwargs : variable length of keyword arguments for the objective function
+
+        Returns:
+            list[array]: the new variable values :math:`x^{(t+1)}`.
+            If single arg is provided, list [array] is replaced by array.
+        """
+
+        self.trainable_args = set()
+
+        for index, arg in enumerate(args):
+            if getattr(arg, "requires_grad", True):
+                self.trainable_args |= {index}
+
+        if self.s is None:
+            # Number of shots per parameter
+            self.s = [
+                np.zeros_like(a, dtype=np.int64) + self.min_shots
+                for i, a in enumerate(args)
+                if i in self.trainable_args
+            ]
+
+        # keep track of the number of shots run
+        s = np.concatenate([i.flatten() for i in self.s])
+        self.max_shots = max(s)
+        self.shots_used += int(2 * np.sum(s))
+
+        # compute the gradient, as well as the variance in the gradient,
+        # using the number of shots determined by the array s.
+        grads, grad_variances = self.compute_grad(objective_fn, args, kwargs)
+        new_args = self.apply_grad(grads, args)
+
+        if self.xi is None:
+            self.chi = [np.zeros_like(g, dtype=np.float64) for g in grads]
+            self.xi = [np.zeros_like(g, dtype=np.float64) for g in grads]
+
+        # running average of the gradient
+        self.chi = [self.mu * c + (1 - self.mu) * g for c, g in zip(self.chi, grads)]
+
+        # running average of the gradient variance
+        self.xi = [self.mu * x + (1 - self.mu) * v for x, v in zip(self.xi, grad_variances)]
+
+        for idx, (c, x) in enumerate(zip(self.chi, self.xi)):
+            xi = x / (1 - self.mu ** (self.k + 1))
+            chi = c / (1 - self.mu ** (self.k + 1))
+
+            # determine the new optimum shots distribution for the next
+            # iteration of the optimizer
+            s = np.ceil(
+                (2 * self.lipschitz * self._stepsize * xi)
+                / (
+                    (2 - self.lipschitz * self._stepsize)
+                    * (chi ** 2 + self.b * (self.mu ** self.k))
+                )
+            )
+
+            # apply an upper and lower bound on the new shot distributions,
+            # to avoid the number of shots reducing below min(2, min_shots),
+            # or growing too significantly.
+            gamma = (
+                (self._stepsize - self.lipschitz * self._stepsize ** 2 / 2) * chi ** 2
+                - xi * self.lipschitz * self._stepsize ** 2 / (2 * s)
+            ) / s
+
+            argmax_gamma = np.unravel_index(np.argmax(gamma), gamma.shape)
+            smax = s[argmax_gamma]
+            self.s[idx] = np.squeeze(np.int64(np.clip(s, min(2, self.min_shots), smax)))

```

Comment:
Yo, I there is a lot going on here :) But you are basically just clipping in a multidim array?

Reply:
Yep! Lines 439-441 I am essentially implementing the following two lines from [Algorithm 2 of the paper](https://arxiv.org/abs/2004.06252):

![image](https://user-images.githubusercontent.com/2959003/111460219-acbdf700-8756-11eb-8ae7-91c26066776f.png)

Reply:
The `np.squeeze` and `np.int64` were added during testing, there were a few edge cases where an extra dimension would be introduced, or `s` would suddenly become a float.